Simplenote/StorageSettings.swift

```diff
@@ -0,0 +1,38 @@
+import Foundation
+
+@objcMembers
+class StorageSettings: NSObject {
+
+    /// URL for the managed object model resource
+    ///
+    let modelURL = Bundle.main.url(forResource: Constants.resourceName, withExtension: Constants.resourceType)!
+
+    /// In app core data storage URL
+    ///
+    let legacyStorageURL = FileManager.default.documentsURL.appendingPathComponent(Constants.sqlFile)
+
+    var legacyStorageExists: Bool {
+        FileManager.default.fileExists(atPath: legacyStorageURL.path)
+    }
+
+    /// URL for core data storage in shared app group documents directory
+    ///
+    let sharedStorageURL = FileManager.default.sharedContainerURL.appendingPathComponent(Constants.sqlFile)
+
+    var sharedStorageExists: Bool {
```

Comment:
I'm a bit torn on the `sharedStorageExists` / `legacyStorageExists` API(s). Both are handy and convenient, but they should probably not live inside `StorageSettings`.

Strictly speaking, the `StorageSettings` tool should not run logic perhaps. But again, i'm a bit torn, because moving elsewhere implies building that elsewhere as well!

Reply:
I didn't love putting them in StorageSettings either.  I moved these checks 2 or 3 times before settling on the StorageSettings class.

I looked at it again and I see a few of options...
1. move `migrationNeeded` to StorageSettings.  Doesn't seem like a natural place for this to live, but then all of the file manager paths and the file exist checks are all stored together in one place
2. move the `sharedStorageExists` and `legacyStorageExists` to SharedStorageMigrator.  This seems like the right place to put them, but because StorageSettings uses these checks we will need to make them static or something to pass back the files exist checks to StorageSettings
3. A bit more radically, we could assume the migration will work and remove the possibility of falling back on the old url.  So instead of :
    var storageURL: URL {
        if legacyStorageExists && !sharedStorageExists {
            return legacyStorageURL
        }
        return sharedStorageURL
    }

we simply use the sharedStorageURL everywhere except the SharedStorageMigrator.  This feels more risky to me cause if a users migration fails they will probably need to reinstall to get the app working again...  But if we assume that the migration will work then the StorageSettings class doesn't need to check for the old URL, only the migrator needs that, so we could move these checks to the migrator and assume the only storage URL needed outside the migrator is the new URL.

TBH, I am nervous about option 3, but it feels like the best choice here.  @eshurakov had asked earlier in the process why we were copying the DB instead of moving it, and I didn't really have a good answer...  Maybe trusting that the migration will work in most cases isn't the worst plan.  What do you think @jleandroperez @eshurakov ?

Reply:
There are several cases that raise questions:

#### 1. DB file migrated and for some reason resulted in unusable/corrupted file.  
I'm not sure how this can happen, but if want to be very careful and guard against it we can do it like this:
```
do {
    coreDataManager = try CoreDataManager(storageURL)
} catch {
    if legacyStorageURLExists {
        coreDataManager = try CoreDataManager(legacyStorageURL)
    } else {
        throw error
    }
}
```

#### 2. DB file failed to migrate.  
We need to be careful here and not accidentally create a new empty database file in new location. I see this happening if we try to initialise core data stack from widgets.
In this scenario widgets will be unusable. This can happen if there is not enough free space on user device. 

If we can determine that we fail because there is not enough space, we can maybe show an error message to the user?

---

#### StorageSettings

The way I would do it keeping the current logic:
* remove `*exists` from settings and move them to `SharedStorageMigrator` as in option 2 suggested by Charle
* return a migration result from `SharedStorageMigrator` and use it to determine which file to use for core data

Something in the following lines:
```
let migrationResult = SharedStorageMigrator(storageSettings: settings).migrate()
switch migrationResult {
case .migrated, .notNeeded:
    // setup core data using new URL
    coreDataManager = try CoreDataManager(storageURL)
case .failed:
    coreDataManager = try CoreDataManager(legacyStorageURL)
}
```

Reply:
Thanks for the suggestions @eshurakov I made some changes that sort of went down the middle of these two options.  I have changed setting up the core data manager to throw, and setup the migrator to return a result if the migration fails.  If the migration fails for any reason the app should attempt to use the legacy storage url, regardless of what the issue was, so in theory would handle a migration failures.

When attempting to load the DB from the new url, if there is an issue it will attempt to load with the old url before failing to setup CD and crashing.  WDYT?